Util/channel/src/main/java/io/deephaven/util/channel/SeekableChannelsProvider.java

```diff
@@ -18,26 +18,69 @@
 public interface SeekableChannelsProvider extends SafeCloseable {
 
     /**
-     * Take the file source path or URI and convert it to a URI object.
+     * Take the file source path or URI string and convert it to a URI object.
      *
      * @param source The file source path or URI
      * @return The URI object
      */
     static URI convertToURI(final String source) {
+        if (source.isEmpty()) {
+            throw new IllegalArgumentException("Cannot convert empty source to URI");
+        }
         final URI uri;
         try {
             uri = new URI(source);
         } catch (final URISyntaxException e) {
             // If the URI is invalid, assume it's a file path
-            return new File(source).toURI();
+            return convertFileToURI(new File(source));
         }
         if (uri.getScheme() == null) {
-            // Need to convert to a "file" URI
-            return new File(source).toURI();
+            // Convert to a "file" URI
+            return convertFileToURI(new File(source));
         }
         return uri;
     }
 
+    /**
+     * Takes a file and convert it to a URI object. We should use this method instead of {@link File#toURI()} because
+     * {@link File#toURI()} internally calls {@link File#isDirectory()}, which can lead to disk access.
+     *
+     * @param file The file
+     * @return The URI object
+     */
+    static URI convertFileToURI(final File file) {
+        String absPath = file.getAbsolutePath();
+        if (absPath.isEmpty()) {
+            throw new IllegalArgumentException("Cannot convert file with empty path to URI: " + file);
+        }
+        if (File.separatorChar != '/') {
+            absPath = absPath.replace(File.separatorChar, '/');
+        }
+        if (absPath.charAt(0) != '/') {
+            absPath = "/" + absPath;
+        }
+        if (absPath.startsWith("//")) {
+            absPath = "//" + absPath;
+        }
+        try {
+            return new URI("file", null, absPath, null);
+        } catch (final URISyntaxException e) {
+            throw new IllegalArgumentException("Failed to convert file to URI: " + file, e);
+        }
+    }
+
+    /**
+     * Takes a path and convert it to a URI object. We should use this method instead of {@link Path#toUri()} because
+     * {@link Path#toUri()} internally does a {@code stat} system call to extract information whether file is a
+     * directory.
+     *
+     * @param path The path
+     * @return The URI object
+     */
+    static URI convertPathToURI(final Path path) {
+        return convertFileToURI(path.toFile());
+    }
+
```

Comment:
I did some basic benchmark tests

```
    @Test
    public void fileToURIProfileTest() {
        final File test = new File("/Users/shivammalhotra/Documents");
        final int numIterations = 5_000_000;
        final double start = System.currentTimeMillis();
        for (int i = 0; i < numIterations; i++) {
            final URI uri = convertFileToURI(test);
        }
        final double end = System.currentTimeMillis();
        System.out.println("Time taken: " + (end - start));


        final double start2 = System.currentTimeMillis();
        for (int i = 0; i < numIterations; i++) {
            final URI uri = test.toURI();
        }
        final double end2 = System.currentTimeMillis();
        System.out.println("Time taken: " + (end2 - start2));

    }

    @Test
    public void pathToURIProfileTest() {
        final Path test = Path.of("/Users/shivammalhotra/Documents");
        final int numIterations = 5_000_000;
        final double start = System.currentTimeMillis();
        for (int i = 0; i < numIterations; i++) {
            final URI uri = convertPathToURI(test);
        }
        final double end = System.currentTimeMillis();
        System.out.println("Time taken: " + (end - start));


        final double start2 = System.currentTimeMillis();
        for (int i = 0; i < numIterations; i++) {
            final URI uri = test.toUri();
        }
        final double end2 = System.currentTimeMillis();
        System.out.println("Time taken: " + (end2 - start2));
    }
```

And in both cases, the time taken by toURI was almost 6x.

Reply:
And you're probably running on a fast, local FS. It can get way worse.